mincblur/mincblur.1
.\" Copyright 1995 Louis Collins, McConnell Brain Imaging Centre,
.\" Montreal Neurological Institute, McGill University.
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies.  The author and McGill University
.\" make no representations about the suitability of this
.\" software for any purpose.  It is provided "as is" without
.\" express or implied warranty.
.\"
.\" $Header: /private-cvsroot/registration/mni_autoreg/mincblur/mincblur.1,v 96.1 2000-01-28 16:21:35 stever Exp $
.\"
.TH MINCBLUR 1

.SH NAME
mincblur - convolve an input volume with a Gaussian blurring kernel.

.SH SYNOPSIS
.B mincblur [<options>] <source> <output_basename>

.SH DESCRIPTION
.I Mincblur
convolves an input volume with a Gaussian blurring kernel of
user-defined width.  Convolution is accomplished by multiplication in
the Fourier domain where the blurring kernel is calculated explicitly
over the whole field.  Mincblur can also calculate the first partial
derivatives and the gradient magnitude volume.

The first command line argument is the name of the 3D input MINC file.
The second argument is the basename for the output file.  The string
"_blur.mnc" is added automatically to the basename to generate the
blurred intensity volume; if a gradient volume is calculated (using
the -gradient option), then "_dxyz.mnc" is added to the basename; and
if partial derivative volumes are calculated (using the -partial
option), then "_dx.mnc", "_dy.mnc", and "_dz.mnc" are added to name
them.  The output volume(s) will have the same size and be of the same
type as the input volume.

Before blurring, the edges of the data volume are apodized (intensity
reduced) to minimize edge effects.  This process can be skipped with
the '-no_apodize' flag.

.SH Specification of blurring kernel.
One of the following options must be used to specify the size of the
blurring kernel: -fwhm, -standarddev, -3Dfwhm, where
.P
.I -fwhm
<val>: Specifies the full-width-half-maximum of the iso-tropic 3D
Gaussian blurring kernel.
.P
.I -standarddev
<val>: Specifies the sigma iso-tropic 3D Gaussian blurring kernel.
.P
.I -3Dfwhm
<valx> <valy> <valz>: Specifies the full-width-half-maximum in the x,
y and z directions of the non-isotropic 3D Gaussian blurring kernel.

.SH Other options
.P
.I -dimensions
<1|2|3>: Number of dimensions to blur. -dim 1 blurs only in the z
direction, useful for simulating thick slices. -dim 2 blurs only in
the xy plane.  -dim 3 blur all 3 dimensions and is the default.
.P
.I -gaussian:
Use a gaussian smoothing kernel (default).
.P
.I -rect:
Use a rect (box) smoothing kernel.
.P
.I -no_apodize:
Do not apodize the data before blurring.
.P
.I -no_clobber:
Do not overwrite output file (default).
.P
.I -clobber:
Overwrite output file.
.P
.I -gradient:
Create the gradient magnitude (_dxyz.mnc) volume as well.
.P
.I -partial:
Create the partial derivative (_dx.mnc, _dy.mnc & _dz.mnc) volumes as well.
.P
In order to create any gradient data, it is necessary to temporarily
store the blurred data in floating point representation.  By default,
the temporary files are written in the current directory. If that
directory is not available, /tmp is used.  Both of these possiblities
can be overridden by providing an environment variable TMPDIR in the
user's environment, whose value is the name of the desired
temporary-file directory.  Of course, the temporary files are removed
by mincblur after processing.
.SH Options for logging progress.
.P
.I -verbose
<val>:
Write verbose messages indicating progress (default = 1).
.P
.I -quiet:
Do not write log messages
.P
.I -debug:
Print out debug info.
.SH Generic options
.P
.I -help:
Print summary of command-line options and abort.

.SH EXAMPLES
1) Blur an input volume with a 6mm fwhm isotropic Gaussian blurring
kernel:

     mincblur -fwhm 6 input.mnc out_6

mincblur will create out_6_blur.mnc.

2) Calculate the blurred and gradient magnitude data:

     mincblur -fwhm 6 -gradient input.mnc out_6

mincblur will create out_6_blur.mnc and out_6_dxyz.mnc

3) Calculate the blurred data, the partial derivative volumes and the
gradient magnitude for the same data:

     mincblur -fwhm 6 -partial input.mnc out_6

mincblur will create out_6_blur.mnc, out_6_dx.mnc, out_6_dy.mnc,
out_6_dz.mnc and out_6_dxyz.mnc

.SH KNOWN BUGS
The temporary files may not be removed properly if mincblur fails
during processing.

.SH AUTHOR
Louis Collins

.SH COPYRIGHT
Copyright (c) 1993-95 by Louis Collins


NEWS
Release 
--------------
* Free memory after usage in mincblur.
* fixed one bug in apodize_data.c for ABS(step) in one direction
* now use VIO_X, VIO_Y, VIO_Z (no more X, Y, Z) in mincblur and
  make it invariant to dimension order

Release 0.99.5
--------------

* input source and volumes are read as DOUBLES.  
* deformation fields are created and manipulated as DOUBLES.
	
Release 0.99.4
--------------

* added -nmi option (normalized mutual information) as an objective
  function for linear registration.
* modeldir is now $prefix/share/mni-models
* fixed a bug for offsets in minctracc/Volume/init_lattice.c
* free memory after usage in minctracc
* file compression (minc2) disabled for temporary files in make_model
* fixed bug in mritoself with same file names in different directories

Release 0.99.3
--------------

* new release to compile on ia64

Release 0.99.2
--------------

* fixed a bug in mritoself and mritotal when piping output to STDOUT

Release 0.99.1
--------------

* fixed a bug in autocrop regarding split_path

Release 0.98v
-------------

* added vladimir fonovs changes to xfmtool to allow .gz grid files


Release 0.98u
-------------

* removed perl libraries that are now part of mni_perllib
* removed c libraries that are now part of MINC proper

Release 0.98t
-------------

* 30% speed increase due to changes in nonlinear fitting code
* Corrected masking in sub-lattice (Patricia Le Nezet and Irina Kezele)


Release 0.98s
-------------

* Removed malloc.h from mincblur to assist build on OSX


Release 0.98r
-------------

* Fixed COV bug (lenezet)
* Fixed make check bug with direction cosines
* added epm-header.in for binary package building


Release 0.98q
-------------

* Removed all reference to public and private syntatic sugar
    private has been replaced with static.  This has caused
	erratic build problems in the past.  Let's see if things
	are better now.
* Removed most Keyword Substitutions in perl programs in favour
    of the perl module FindBin::Bin this module is part of the
	base perl distribution


Release 0.98p
-------------

* Fixed problems with Direction cosines of input volumes
* Changed from volume_stats to mincstats in minc_utilities.pl
   this will mean that we no longer have a dependency on N3.Z


Release 0.98j
-------------

* Build process changed: third-party libraries will no longer be
  searched for in /usr/local and ${prefix} by default.  Specify
  the directories to search using --with-build-path.


Release 0.98i
-------------

This is a bug-fix release; no user-visible changes.


Release 0.98h
-------------

* mritotal handles nonlinear as well as linear fitting.
* Build process revamped: supports non source-directory builds; no
  environment variables are required to specify library locations.
* Fixed bug that caused PROCRUSTES transformations to optimize only
  the x-scale, rather than a global scale (as LSQ7 does).
* Generated files contain the long version string in their comment section.


Release 0.98g
-------------

* Minctracc is built with both linear and non-linear registration.


Release 0.98f
-------------

* version 0.98e did not give the same registration results as 0.98d when
  using masks due to an oversight in the voxel coordinate manipulation
  in the mask volumes - an important problem for using mritotal.  
  This has been corrected in 0.98f.


Release 0.98e
-------------

* feature volumes are supported to be able to register sets (like T1, T2 and
  classified data) of volumes from a source data set to match corresponding
  features of a target.
* the minctracc registration engine is approximately 2.5 times faster for
  resampling data during evaluation of the optimization function.  This
  results in an overall reduction in execution time by a factor of 2.


Release 0.98
------------

* added mritoself to list of programs to install (in perl/Makefile)
* slight tweak to test of compiler's ANSIness in configure script
* fixed some subtle but important bugs in mritotal, autocrop, and mritoself
* more enhancements to the Perl libraries
* removed an unused option (-Lvv) from mincblur's argument table
* removed ParseArgs module from the distribution -- Getopt::Tabular 
  (available separately) is now required
* changed mritotal's algorithm for trimming volumes after blurring
* added '-bbox' feature to autocrop
bug fixes:
release 0.98b, 0.98c: fixed two bugs in mritoself


Release 0.97
------------

* fixed those bugs in mritotal
* enhanced configure script for portability and flexibility
* added the TESTING document to explain testing the package in detail


Release 0.96
------------

* inexplicable jump of four micro-versions instead of one
* introduced some buggy features in mritotal


Release 0.92
------------

* the implementation of the mutual information objective function in
  minctracc has been greatly improved
* many portability problems have been fixed
* some real bugs (particularly in mincblur) have been fixed
* most of the Perl libraries have been greatly updated to take
  better advantage of Perl 5
* added -blur/-noblur and -firstobj options to mritotal (experimental --
  not used by any of the standard fitting procedures)


Release 0.91
------------

Lost in the mists of time...


Release 0.9
-----------

First public release.


mincblur/mincblur.c
/* ----------------------------- MNI Header -----------------------------------
   @NAME       : mincblur.c
   @INPUT      : -name of file corresponding to minc volume data
                 -basename for output file
                 -size of blurring kernel
   
   @OUTPUT     : one or more of five volumes, depending on the command-line args:

        1 - blurred volume - volume blurred by gaussian kernel (sigma = k/2.36)
        2 - d/dx           - derivative along x of blurred volume.
        3 - d/dy           - derivative along y of blurred volume.
        4 - d/dz           - derivative along z of blurred volume.
        5 - | d^3/dxdydz | = sqrt ( (d/dx)^2 + (d/dy)^2 + (d/dz)^2 ) 
                           i.e. the derivative magnitude of the blurred volume.

   @RETURNS    : TRUE if ok, VIO_ERROR if error.

   @DESCRIPTION: This program will read in a volumetric dataset in
                 .mnc format.  This file is assumed to contain rectangular 
                 voxels.

   @METHOD     : The blurred volume calculated in the FOURIER domain by 
                 multiplying the FT(data) by the FT(Gaussian).

           blurred vol = blur_z ( blur_y( blur_x(data) ) );

   the gradient volumes are calculated also in the fourier domain,
   for each direction,
                      -1
        d/dx data = FT  ( is *  FT(blurred_data) ) ; where i = imag, s = freq var.
                      -1
        d/dy data = FT  ( is *  FT(blurred_data) )
                      -1
        d/dz data = FT  ( is *  FT(blurred_data) )

   the gradient magnitude is calculated so that at each voxel of the gradient volume,
   the intensity is equal to:

       grad_mag = sqrt ( (d/dx)^2 + (d/dy)^2 + (d/dz)^2 ) 

   @GLOBALS    : char *prog_name - stores the name of the program.
                 int  debug      - prints out debugging info   
                 int  verbose    - prints out running info

   @COPYRIGHT  :
              Copyright 1995 Louis Collins, McConnell Brain Imaging Centre, 
              Montreal Neurological Institute, McGill University.
              Permission to use, copy, modify, and distribute this
              software and its documentation for any purpose and without
              fee is hereby granted, provided that the above copyright
              notice appear in all copies.  The author and McGill University
              make no representations about the suitability of this
              software for any purpose.  It is provided "as is" without
              express or implied warranty.
   @CREATED    : January 25, 1992 louis collins (Original using .iff files)
   @MODIFIED   : $Log: mincblur.c,v $
   @MODIFIED   : Revision 96.7  2009-07-23 22:34:00  claude
   @MODIFIED   : cleanup in mincblur for VIO_X, VIO_Y, VIO_Z
   @MODIFIED   :
   @MODIFIED   : Revision 96.6  2009/06/05 20:49:52  claude
   @MODIFIED   : Free memory after usage in mincblur
   @MODIFIED   :
   @MODIFIED   : Revision 96.5  2009/02/13 04:14:40  rotor
   @MODIFIED   :  * small updated to arguments of mincblur
   @MODIFIED   :
   @MODIFIED   : Revision 96.4  2006/11/28 09:12:21  rotor
   @MODIFIED   :  * fixes to allow clean compile against minc 2.0
   @MODIFIED   :
   @MODIFIED   : Revision 96.3  2005/07/20 20:45:39  rotor
   @MODIFIED   :     * Complete rewrite of the autoconf stuff (configure.in -> configure.am)
   @MODIFIED   :     * Many changes to includes of files (float.h, limits.h, etc)
   @MODIFIED   :     * Removed old VOLUME_IO cruft #defines
   @MODIFIED   :     * Fixed up all Makefile.am's in subdirs
   @MODIFIED   :     * Removed all things in Proglib that are now part of MINC proper
   @MODIFIED   :     * Still working on fixing up perl subdirectory - removing mni_perllib
   @MODIFIED   :
   @MODIFIED   : Revision 96.2  2000/02/20 04:01:00  stever
   @MODIFIED   : * use new history_string() function to generate history strings
   @MODIFIED   :   when outputting MNI files (.mnc, .xfm)
   @MODIFIED   : * removed unused vax routines from Proglib
   @MODIFIED   : * tuned configure script; CPPFLAGS and LDFLAGS are now left alone,
   @MODIFIED   :   for the installer to use
   @MODIFIED   :
   @MODIFIED   : Revision 96.1  2000/01/27 18:03:52  louis
   @MODIFIED   : final checkin before switch to CVS
   @MODIFIED   :
 * Revision 96.0  1996/08/21  18:22:24  louis
 * Release of MNI_AutoReg version 0.96
 *
 * Revision 9.6  1996/08/21  18:22:20  louis
 * Pre-release
 *
 * Revision 9.5  1996/08/12  14:16:28  louis
 * Never released version 0.95
 *
 * Revision 1.14  1996/08/12  14:16:24  louis
 * Pre-release
 *
 * Revision 1.13  1995/09/18  09:02:42  collins
 * new functional version of mincblur with new and improved default behavior.
 * By default, only the blurred volume is created. If you want the gradient
 * magnitude volumes, you have to ask for it (-gradient).
 *
 * The temporary files (corresponding to the REAL data, and partial derivatives)
 * are removed by mincblur, so now it runs cleanly.  Unfortunately, these files
 * are _not_ deleted if mincblur fails, or is stopped.  I will have to use
 * unlink for this, but its a bit to much to do right now, since I would have
 * to change the way files are dealt with in gradient_volume.c, blur_volume.c
 * and gradmag_volume.c.
 *
 * Also, it is possible to keep the partial derivative volumes (-partial).
 *
 * this version is in mni_reg-0.1j
 *
 * Revision 1.12  1995/09/18  06:45:42  collins
 * this file is a working version of mincblur.  All references to numerical
 * recipes routines have been removed.  This version is included in the
 * package mni_reg-0.1i
 *
   Wed Jun 23 09:04:34 EST 1993 Louis Collins
        rewrite using mnc files and David Macdonald's libmni.a
   ---------------------------------------------------------------------------- */
#ifndef lint
static char rcsid[]="$Header: /static-cvsroot/registration/mni_autoreg/mincblur/mincblur.c,v 96.7 2009-07-23 22:34:00 claude Exp $";
#endif

#include <config.h>
#include <float.h>
#include <volume_io.h>
#include <Proglib.h>
#include <ParseArgv.h>
#include <minc.h>
#include "kernel.h"
#include "mincblur.h"

/* confiscated from minctracc.c */
void get_volume_XYZV_indices(VIO_Volume data, int xyzv[]) {

  int     axis, i, vol_dims;
  char ** data_dim_names;

  vol_dims       = get_volume_n_dimensions(data);
  data_dim_names = get_volume_dimension_names(data);

  for(i=0; i<VIO_N_DIMENSIONS+1; i++) xyzv[i] = -1;
  for(i=0; i<vol_dims; i++) {
    if (convert_dim_name_to_spatial_axis(data_dim_names[i], &axis )) {
      xyzv[axis] = i;
    } else {     /* not a spatial axis */
      xyzv[VIO_Z+1] = i;
    }
  }

  delete_dimension_names(data, data_dim_names);
}
void XYZV_to_RCSV_indices(VIO_Volume data, int xyzv[], int rcsv[]) {
  int i;
  
  for(i=0; i<VIO_N_DIMENSIONS+1; i++) rcsv[i] = xyzv[i];
  if ( equal_strings(slice_dim, "xspace") ){
    rcsv[0] = xyzv[2];
    rcsv[1] = xyzv[1];
    rcsv[2] = xyzv[0];
  }
  else if ( equal_strings(slice_dim, "yspace") ){
    rcsv[0] = xyzv[2];
    rcsv[1] = xyzv[0];
    rcsv[2] = xyzv[1];
  }
  else if ( equal_strings(slice_dim, "zspace") ){
    rcsv[0] = xyzv[0];
    rcsv[1] = xyzv[1];
    rcsv[2] = xyzv[2];
  }
  else{
    print_error_and_line_num ("Could not interpret slice dimension\n", __FILE__, __LINE__);
  }
}

int main (int argc, char *argv[] )
{   
  
  FILE 
    *ofd, *reals_fp;
 
  char 
    *infilename,
    *partials_name,
    *output_basename,
    *temp_basename,
    tempname[1024],
    reals_filename[1024];
  VIO_Status 
    status;
  
  VIO_Volume
    data;
  VIO_Real
    min_value, max_value,
    step[3];
  int
    n_dimensions,
    i,
    sizes[3],
    xyzv[VIO_MAX_DIMENSIONS], // indices of x, y, and z dimensions
    rcsv[VIO_MAX_DIMENSIONS]; // indices of row, column, slice dimensions
  char *history;
  char *tname;

  /* set default values */
  prog_name            = argv[0];
  apodize_data_flg     = TRUE;
  clobber_flag         = FALSE;
  verbose              = TRUE;
  debug                = FALSE;
  do_gradient_flag     = FALSE;
  do_partials_flag     = FALSE;
  infilename           = (char *)NULL;
  output_basename      = (char *)NULL;
  temp_basename        = (char *)NULL;
  partials_name        = (char *)NULL;
  ofd                  = (FILE *)NULL;
  reals_fp             = (FILE *)NULL;
  dimensions           = 3;
  kernel_type          = KERN_GAUSSIAN;
                                /* init kernel size */
  standard             =  0.0;
  fwhm                 =  0.0;
  for(i=0; i<3; i++){
    fwhm_3D[i] = -DBL_MAX;
  } 
  
  history = history_string(argc, argv);


   /******************************************************************************/
  /* Call ParseArgv to interpret all command line args                          */

  if (ParseArgv(&argc, argv, argTable, 0) || (argc!=3)) {
    (void) fprintf(stderr, 
                   "\nUsage: %s [<options>] <inputfile> <output_basename>\n", 
                   prog_name);
    (void) fprintf(stderr,"       %s [-help]\n\n", prog_name);
    exit(EXIT_FAILURE);
  }

  /******************************************************************************/
  /* find the size of the blurring kernel, from one of -std, -fwhm, -fwhm3d     */

  if (standard==0.0 && fwhm==0.0 && 
      fwhm_3D[0]==-DBL_MAX && fwhm_3D[1]==-DBL_MAX && fwhm_3D[2]==-DBL_MAX ) {
    print_error_and_line_num ("Must specify either -fwhm, -3D_fwhm or -standard on command line.\n", 
                 __FILE__, __LINE__);
  }

  if (fwhm==0.0) fwhm=standard*2.35;
  
  if (fwhm !=0.0 ) {
    for(i=0; i<3; i++) fwhm_3D[i] = fwhm;
  };                                

  /******************************************************************************/
  /*                   set up necessary file names                              */
  /******************************************************************************/
  infilename      = argv[1];        
  output_basename = argv[2]; 

                                /* check to see if the output file can be written */

  if (!clobber_flag ) {
    tname = malloc(strlen(output_basename)+strlen("_blur.mnc")+2 * sizeof(char*));
    strcpy(tname,output_basename);
    strcat(tname,"_blur.mnc");
    if( file_exists(tname)) {
      print ("File %s exists. -- %d\n", tname);
      print ("Use -clobber to overwrite.\n");
      free(tname);
      return VIO_ERROR;
    }
    free(tname);
  }

  status = open_file( output_basename , WRITE_FILE, BINARY_FORMAT,  &ofd );
  if ( status != VIO_OK ) 
    print_error_and_line_num ("filename `%s' cannot be opened.", 
                              __FILE__, __LINE__, output_basename);
  status = close_file(ofd);
  remove(output_basename);   

       /* if any gradient data is needed, then we have to save the blurred
          volume in float representation, otherwise quantization errors can
          mess up the derivatives.  So get temporary file to save 'real data' */

  if ((do_partials_flag || do_gradient_flag)) {

    temp_basename = tempnam(NULL,"mblur");
    
    if (temp_basename == (char *)NULL) {
      print_error_and_line_num ("Can't build temporary filename.", 
                                __FILE__, __LINE__);
    }

    sprintf(reals_filename,"%s_reals.raw",temp_basename);

    status = open_file( reals_filename, WRITE_FILE, BINARY_FORMAT,  &reals_fp );
    
    if (status != VIO_OK) {
      print_error_and_line_num ("Temporary file to save blurred volume cannot be opened.", 
                                __FILE__, __LINE__);
    }

  }


  /******************************************************************************/
  /*             create blurred volume first                                    */
  /******************************************************************************/

  status = input_volume(infilename, VIO_N_DIMENSIONS, 
                        get_default_dim_names( VIO_N_DIMENSIONS ),
                        NC_UNSPECIFIED, FALSE, 0.0, 0.0, TRUE, 
                        &data, (minc_input_options *)NULL);
  if ( status != VIO_OK )
    print_error_and_line_num("problems reading `%s'.\n",__FILE__, __LINE__,infilename);


  /* Get indices of x/y/z dimensions */
  get_volume_XYZV_indices( data, xyzv );

  /* Convert to row/col/slice indicies */
  XYZV_to_RCSV_indices( data, xyzv, rcsv );

  if (debug) {
    get_volume_sizes(data, sizes);
    get_volume_separations(data, step);
    printf ( "===== Debugging information from %s =====\n", prog_name);
    printf ( "Data filename     = %s\n", infilename);
    printf ( "Output basename   = %s\n", output_basename);
    printf ( "Input volume      = %3d cols by %3d rows by %d slices\n",
            sizes[xyzv[VIO_X]], sizes[xyzv[VIO_Y]], sizes[xyzv[VIO_Z]]);
    printf ( "Input voxels are  = %8.3f %8.3f %8.3f\n", 
            step[xyzv[VIO_X]], step[xyzv[VIO_Y]], step[xyzv[VIO_Z]]);
    get_volume_real_range(data,&min_value, &max_value);
    printf ( "min/max value     = %8.3f %8.3f\n", min_value, max_value);
  }
  
  n_dimensions = get_volume_n_dimensions (data);
  if (n_dimensions!=3) {
    print_error_and_line_num ("File %s has %d dimensions.  Only 3 dims supported.", 
                              __FILE__, __LINE__, infilename, n_dimensions);
  }
  
                                /* apodize data if needed */
  if (apodize_data_flg) {
    if (debug) print ("Apodizing data at (%f,%f) (%f,%f) (%f,%f)\n",
                      fwhm_3D[0], fwhm_3D[0], fwhm_3D[1], fwhm_3D[1], fwhm_3D[2], fwhm_3D[2] );
    apodize_data(data, xyzv, fwhm_3D[0], fwhm_3D[0], fwhm_3D[1], fwhm_3D[1], fwhm_3D[2], fwhm_3D[2] );
  }
 
                                /* now _BLUR_ the DATA! */
  status = blur3D_volume(data, rcsv,
                         fwhm_3D,
                         infilename,
                         output_basename,
                         reals_fp,
                         dimensions,kernel_type,history);

  /******************************************************************************/
  /*             calculate d/dx,  d/dy and d/dz volumes                         */
  /******************************************************************************/

  if ((do_partials_flag || do_gradient_flag)) {

                                /* reopen REALS file for READ */
    status = close_file( reals_fp );
    if (status!=VIO_OK)
      print_error_and_line_num("Error closing <%s>.",__FILE__, __LINE__, reals_filename);
    
    status = open_file( reals_filename ,READ_FILE, BINARY_FORMAT,  &reals_fp );
    if (status!=VIO_OK)
      print_error_and_line_num("Error opening <%s>.",__FILE__, __LINE__, reals_filename);
      

    if (do_partials_flag) 
      partials_name = output_basename;
    else 
      partials_name = temp_basename;


    status = gradient3D_volume(reals_fp, data, rcsv, infilename, partials_name, dimensions,
                               history, FALSE);
    if (status!=VIO_OK)
      print_error_and_line_num("Can't calculate the gradient volumes.",__FILE__, __LINE__);


                                /* close and delete the REALS temp data file */
    status = close_file( reals_fp );
    if (status!=VIO_OK)
      print_error_and_line_num("Error closing <%s>.",__FILE__, __LINE__, reals_filename);

    remove_file( reals_filename );
    

    /*************************************************************************/
    /*           calculate and save the gradient magnitude volume            */
    /*************************************************************************/
    
    calc_gradient_magnitude(partials_name, output_basename, history, 
                            &min_value, &max_value);
      
                                /* remove the partial derivs, unless
                                   the user specifically wants to keep
                                   them */
    if (!do_partials_flag) {
      sprintf (tempname,"%s_dx.mnc",partials_name); remove_file(tempname);
      sprintf (tempname,"%s_dy.mnc",partials_name); remove_file(tempname);
      sprintf (tempname,"%s_dz.mnc",partials_name); remove_file(tempname);
    }

  }

  delete_volume( data );
  if( history ) free( history );
  if( temp_basename ) free( temp_basename );

  return(status);
   
}


